### PR TITLE
HC entropy

### LINKED ISSUE #3926: HC: pin contract environment on top of older block
At the moment a leader can cherry pick which calls to the staking contract to accept and thus - effectively manipulate the next-next leader. Instead we shall make the election based on a state in the past. How many keyblocks in the past shall we go back would be configurable.

### LINKED ISSUE #3927: HC: parent chain - get latest hash
The parent chains' latest block must be passed all the way to the child chain's smart contract. There it should be used as a source of entropy.

### LINKED ISSUE #3962: HC: conductor error message
We observed a lot of those on HC testnet:
```
2022-07-18 13:23:52.114 [error] <0.1702.0>@aec_conductor:handle_key_block_candidate_reply:1243 Creation of key block candidate failed: no_candidate
```

Those do make sense in PoW context but are not appropriate in PoS context. Investigate why those errors happen. 

Until this is resolved, HC logging is disabled.

### LINKED ISSUE #3974: HC remove default json files
Remove HC related default files in `data/aecore/.ceres` - they should be generated instead.

### PR DESCRIPTION
**Pull request by velzevur**
Fixes #3927 
Fixes #3926 
Fixes #3974 
Fixes #3962 

This PR:
* introduces entropy derived from the parent chain
* includes PC caching mechanism
* the MainStaking contract is split into two: a staking contract and an election contract (that could be specific for different pos or hc chains)
* fixes some small bugs
* refactors the parent chain connector and the Aeternity API consuming module

What is left:

- [x] in `aehttp_stake_contract_SUITE` a test (`hc.mine_and_sync`) fails
- [x] delete now unnecessary files in `data/aecore/.ceres` (those are already generated)
- [x] release notes :)

### CODE REVIEW

File: apps/aecore/src/aec_conductor.erl [hunk 1]
```diff
@@ -1009,6 +1010,7 @@ start_block_production_(#state{key_block_candidates = [{_, #candidate{top_hash =
     epoch_mining:info("Key block candidate for old top hash; regenerating"),
     create_key_block_candidate(State);
 start_block_production_(#state{mode = local_pow, key_block_candidates = [{ForSealing, Candidate} | Candidates]} = State) ->
+    lager:info("ASDF start block generation", []),
```
Comment (seanhinde) [inline apps/aecore/src/aec_conductor.erl:1013]: ASDF?

**Review by seanhinde — DISMISSED** [review]
Lovely stuff

File: test/contracts/PoSElection.aes [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  entrypoint leader() =
+    state.leader
+
+  entrypoint get_state() : get_state_response =
+    { main_staking_ct = state.main_staking_ct,
+      entropy         = state.entropy,
+      leader          = state.leader
+      }
+
+
```
Comment (uwiger) [inline test/contracts/PoSElection.aes:51]:
```suggestion
```

File: apps/aehttp/test/aehttp_stake_contract_SUITE.erl [hunk 3]
```diff
@@ -678,3 +614,194 @@ calc_rewards(RewardForHeight) ->
...
+                            ?INITIAL_STAKE, BRIPub, 1),
+    Call10 =
+        contract_call_spec(SCId, ?STAKING_CONTRACT,
+                            "set_validator_description",
+                            ["\"This validator is offline. She can never become a leader. She has no name set. She is receiving the BRI rewards\""],
+                            0, BRIPub, 2),
+    %% keep the BRI offline
+    AllCalls =  [Call1, Call2, Call3, Call4, Call5, Call6,
+                    Call7, Call8, Call9,
+                    Call10],
```
Comment (uwiger) [inline apps/aehttp/test/aehttp_stake_contract_SUITE.erl:711]:
```suggestion
    AllCalls =  [Call1, Call2, Call3, Call4, Call5, Call6,
		 Call7, Call8, Call9,
		 Call10],
```

**Review by uwiger — APPROVED** [review]

**Review by fabiankrol — APPROVED** [review]

File: apps/aecore/src/aec_parent_connector.erl [hunk 4]
```diff
@@ -73,6 +83,23 @@ stop() ->
...
     gen_server:call(?SERVER, trigger_fetch).
 
+request_block_by_hash(Hash) ->
+    gen_server:cast(?SERVER, {request_block_by_hash, Hash}).
+
+request_block_by_height(Height) ->
+    gen_server:cast(?SERVER, {request_block_by_height, Height}).
+
+request_top() ->
+    ?SERVER ! check_parent.
```
[the remark this answers is not in the document]
Comment (velzevur) [reply apps/aecore/src/aec_parent_connector.erl:93]: reusing the existing info handling

**Review by fabiankrol — APPROVED** [review]

**Review by uwiger — APPROVED** [review]

**Pull request merged**